python/paddle/fluid/contrib/int8_inference/utility.py

```diff
@@ -0,0 +1,711 @@
+        '''
+        Update the convolution scale information.
+        '''
+        func = self.__get_optimal_scaling_factor if self.algo == 'KL' else np.max
+
+        for i in self._conv_op_index[1:]:
+            weights_var_name = self.program.current_block().ops[i].input(
+                'Filter')[0]
+            input_var_name = self.program.current_block().ops[i].input('Input')[
+                0]
+            output_var_name = self.program.current_block().ops[i].output(
+                'Output')[0]
+            self._output_program.current_block().ops[i]._set_attr(
+                "Scale_weights", self._var_max_value_map[weights_var_name])
+            self._output_program.current_block().ops[i]._set_attr(
+                "Scale_in", self._var_max_range[input_var_name] /
+                func(self._var_max_value_map[input_var_name]))
+
+            self._output_program.current_block().ops[i]._set_attr(
+                "Scale_out", self._var_max_range[output_var_name] /
+                func(self._var_max_value_map[output_var_name]))
+            if self._output_program.current_block().ops[i].desc.input(
+                    "ResidualData"):
+                residual_var_name = self._output_program.current_block().ops[
+                    i].desc.input("ResidualData")[0]
+                self._output_program.current_block().ops[i]._set_attr(
+                    "Scale_in_eltwise", self._var_max_range[residual_var_name] /
+                    func(self._var_max_value_map[residual_var_name]))
+
+    def __sampling(self, sampling_data):
+        '''
+        Sampling the variables data range.
+        '''
+        for i in self.program.list_vars():
+            if i.name not in self._weights_var_name + self._conv_input_var_name + self._conv_output_var_name + self._residual_input_var_name:
+                continue
+
+            if i.name in self._weights_var_name:
+                max_value = []
+                data = sampling_data[i.name][0]
+                for j in range(data.shape[0]):
+                    if not self._is_close(float(np.max(np.abs(data[j]))), 0.0):
+                        max_value.append(Calibrator.s8_max /
+                                         float(np.max(np.abs(data[j]))))
+                    else:
+                        max_value.append(0.0)
+                max_range = Calibrator.s8_max
+            else:
+                if i.name in self._conv_output_var_name:
+                    cur_op = self.program.current_block().ops[
+                        self._conv_output_var_name.index(i.name) + 2]
+                    if cur_op.has_attr('fuse_relu') and cur_op.attr(
+                            'fuse_relu'):
+                        max_range = Calibrator.u8_max
+                        self._u8_output_var.append(i.name)
+                    else:
+                        max_range = Calibrator.s8_max
+                        self._s8_output_var.append(i.name)
+                else:
```

Comment:
Is this else case not weight, not output, so for inputs and residual_inputs? Where are the biases?

Reply:
As the Scale(Bias) = Scale(Bias)* Scale(Weights),  we don't need to acquire the bias scale separately.

Reply:
@hshen14 would u please give a more detail explanation on this question?

Reply:
Guoming is right.